File: vkatz-lib/src/main/java/by/vkatz/widgets/FlowLayout.java
package by.vkatz.widgets;

import android.content.Context;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

import by.vkatz.R;

/**
 * Created by vKatz on 12.10.2015.
 */
public class FlowLayout extends ViewGroup {
    private static final int GRAVITY_CENTER = 0;
    private static final int GRAVITY_TOP = 1;
    private static final int GRAVITY_BOT = 2;

    private int offsetVertical, offsetHorizontal, lineGravity, fixedLineHeight;
    private ArrayList<Integer> lineSizes;

    public FlowLayout(Context context) {
        this(context, null);
    }

    public FlowLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public FlowLayout(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        lineSizes = new ArrayList<>();
        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.FlowLayout, defStyle, 0);
        offsetHorizontal = a.getDimensionPixelSize(R.styleable.FlowLayout_offsetHorizontal, 0);
        offsetVertical = a.getDimensionPixelSize(R.styleable.FlowLayout_offsetVertical, 0);
        fixedLineHeight = a.getDimensionPixelSize(R.styleable.FlowLayout_lineHeight, -2);
        lineGravity = a.getInteger(R.styleable.FlowLayout_lineGravity, GRAVITY_CENTER);
        a.recycle();
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        int childLeft = getPaddingLeft();
        int childTop = getPaddingTop();
        int lineHeight = fixedLineHeight > 0 ? fixedLineHeight : 0;
        int myWidth = r - l;
        int line = 0;
        for (int i = 0; i < getChildCount(); i++) {
            final View child = getChildAt(i);
            if (child.getVisibility() == View.GONE) continue;
            int childWidth = child.getMeasuredWidth();
            int childHeight = child.getMeasuredHeight();
            if (fixedLineHeight <= 0) lineHeight = Math.max(childHeight, lineHeight);
            if (childWidth + childLeft + getPaddingRight() > myWidth) {
                childLeft = getPaddingLeft();
                childTop += offsetVertical + lineHeight;
                if (fixedLineHeight <= 0) lineHeight = childHeight;
                line++;
            }
            int lineSize = lineSizes.get(line);
            int offset = 0;
            if (lineGravity == GRAVITY_TOP) offset = 0;
            else if (lineGravity == GRAVITY_CENTER) offset = (lineSize - childHeight) / 2;
            else if (lineGravity == GRAVITY_BOT) offset = lineSize - childHeight;
            child.layout(childLeft, offset + childTop, childLeft + childWidth, offset + childTop + childHeight);
            childLeft += childWidth + offsetHorizontal;
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        int childLeft = getPaddingLeft();
        int childTop = getPaddingTop();
        int lineHeight = fixedLineHeight > 0 ? fixedLineHeight : 0;
        int myWidth = resolveSize(100, widthMeasureSpec);
        int wantedHeight = 0;
        lineSizes.clear();
        for (int i = 0; i < getChildCount(); i++) {
            final View child = getChildAt(i);
            if (child.getVisibility() == View.GONE) continue;
            child.measure(
                    getChildMeasureSpec(widthMeasureSpec, getPaddingLeft() + getPaddingRight(), child.getLayoutParams().width),
                    getChildMeasureSpec(heightMeasureSpec, getPaddingTop() + getPaddingBottom(), child.getLayoutParams().height));
            int childWidth = child.getMeasuredWidth();
            int childHeight = child.getMeasuredHeight();
            if (fixedLineHeight <= 0) lineHeight = Math.max(childHeight, lineHeight);
            if (childWidth + childLeft + getPaddingRight() > myWidth) {
                childLeft = getPaddingLeft();
                childTop += offsetVertical + lineHeight;
                if (fixedLineHeight <= 0) lineHeight = childHeight;
                lineSizes.add(lineHeight);
            }
            childLeft += childWidth + offsetHorizontal;
        }
        lineSizes.add(lineHeight);
        if (getChildCount() == 0) lineHeight = 0;
        wantedHeight += childTop + lineHeight + getPaddingBottom();
        setMeasuredDimension(myWidth, resolveSize(wantedHeight, heightMeasureSpec));
    }
}

File: katzext/src/main/kotlin/by/vkatz/katzext/widgets/FlowLayout.kt
package by.vkatz.katzext.widgets

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import by.vkatz.katzext.R
import java.util.*

open class FlowLayout : ExtendRelativeLayout {

    private var offsetVertical: Int = 0
    private var offsetHorizontal: Int = 0
    private var lineGravity: Int = 0
    private var fixedLineHeight: Int = 0
    private lateinit var lineSizes: ArrayList<Int>

    constructor(context: Context) : this(context, null)

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        init(context, attrs, defStyle)
    }

    private fun init(context: Context, attrs: AttributeSet?, defStyle: Int) {
        lineSizes = ArrayList()
        val a = context.obtainStyledAttributes(attrs, R.styleable.FlowLayout, defStyle, 0)
        offsetHorizontal = a.getDimensionPixelSize(R.styleable.FlowLayout_offsetHorizontal, 0)
        offsetVertical = a.getDimensionPixelSize(R.styleable.FlowLayout_offsetVertical, 0)
        fixedLineHeight = a.getDimensionPixelSize(R.styleable.FlowLayout_lineHeight, -2)
        lineGravity = a.getInteger(R.styleable.FlowLayout_lineGravity, GRAVITY_CENTER)
        a.recycle()
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        var childLeft = paddingLeft
        var childTop = paddingTop
        var lineHeight = if (fixedLineHeight > 0) fixedLineHeight else 0
        val myWidth = r - l
        var line = 0
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child.visibility == View.GONE) continue
            val childWidth = child.measuredWidth
            val childHeight = child.measuredHeight
            if (fixedLineHeight <= 0) lineHeight = Math.max(childHeight, lineHeight)
            if (childWidth + childLeft + paddingRight > myWidth) {
                childLeft = paddingLeft
                childTop += offsetVertical + lineHeight
                if (fixedLineHeight <= 0) lineHeight = childHeight
                line++
            }
            val lineSize = lineSizes[line]
            var offset = 0
            when (lineGravity) {
                GRAVITY_TOP -> offset = 0
                GRAVITY_CENTER -> offset = (lineSize - childHeight) / 2
                GRAVITY_BOT -> offset = lineSize - childHeight
            }
            child.layout(childLeft, offset + childTop, childLeft + childWidth, offset + childTop + childHeight)
            childLeft += childWidth + offsetHorizontal
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        var childLeft = paddingLeft
        var childTop = paddingTop
        var lineHeight = if (fixedLineHeight > 0) fixedLineHeight else 0
        val myWidth = View.resolveSize(100, widthMeasureSpec)
        var wantedHeight = 0
        lineSizes.clear()
        for (i in 0 until childCount) {
            val child = getChildAt(i)
            if (child.visibility == View.GONE) continue
            child.measure(
                    ViewGroup.getChildMeasureSpec(widthMeasureSpec, paddingLeft + paddingRight, child.layoutParams.width),
                    ViewGroup.getChildMeasureSpec(heightMeasureSpec, paddingTop + paddingBottom, child.layoutParams.height))
            val childWidth = child.measuredWidth
            val childHeight = child.measuredHeight
            if (fixedLineHeight <= 0) lineHeight = Math.max(childHeight, lineHeight)
            if (childWidth + childLeft + paddingRight > myWidth) {
                childLeft = paddingLeft
                childTop += offsetVertical + lineHeight
                if (fixedLineHeight <= 0) lineHeight = childHeight
                lineSizes.add(lineHeight)
            }
            childLeft += childWidth + offsetHorizontal
        }
        lineSizes.add(lineHeight)
        if (childCount == 0) lineHeight = 0
        wantedHeight += childTop + lineHeight + paddingBottom
        setMeasuredDimension(myWidth, View.resolveSize(wantedHeight, heightMeasureSpec))
    }

    companion object {
        private val GRAVITY_CENTER = 0
        private val GRAVITY_TOP = 1
        private val GRAVITY_BOT = 2
    }
}